Add EAM-generated dust deposition capability to ocean

maltrud opened the pull request:
Currently, the atmospheric dust and iron deposition to the surface ocean is specified to be a monthly climatology that has been used for all E3SM ocean BGC cases for the past decade.  This PR now allows the use of a climatology generated by EAM from archival simulations with prognostic dust deposition and time-variable solubility for iron in dust, which requires extra fields to be read in.  In addition, the wet and dry dust flux that EAM sends through the coupler can now also be used by the ocean in fully coupled runs so only the solubility is required to be read in from a file.

[NML]
[non-BFB] - only for cases with ocean or sea-ice BGC

maltrud committed:
add capability to use prognostic dust/Fe from atm

Commit:
introduce config_use_ecosysTracers_atm_dust_from_coupler

njeffery commented on components/mpas-ocean/src/shared/mpas_ocn_tracer_ecosys.F:
Good idea! Let's coordinate with Yan and find out where the EAM constant is.

njeffery reviewed (approved):
Approved after running a fully coupled test and verifying ocean and sea ice dust/iron concentrations.  The iron molar mass is the same in ocean/seaice.  Would be good to make this a shared constant across all components.

mark-petersen reviewed (changes requested):
I tested MPAS-O standalone with the compass nightly suite on perlmutter with gnu and it passes compile and run, but that does not compile in the coupled code.

I also ran a smoke test with E3SM using 
```
./create_test SMS_Ld3.T62_oQU240.GMPAS-NYF.pm-cpu_gnu -p m4572 -q debug --walltime 00:30:00
```
and it has a compile error at
```
/pscratch/sd/m/mpeterse/e3sm_scratch/pm-cpu/SMS_Ld3.T62_oQU240.GMPAS-NYF.pm-cpu_gnu.20251003_095213_nlxhjj/bld/cmake-bld/core_ocean/shared/mpas_ocn_tracer_ecosys.f90:2868:69:

 2868 |              if (.not. config_use_ecosysTracers_atm_dust_from_coupler) then
      |                                                                     1
Error: Symbol ‘config_use_ecosystracers_atm_dust_from_coupler’ at (1) has no IMPLICIT type
/pscratch/sd/m/mpeterse/e3sm_scratch/pm-cpu/SMS_Ld3.T62_oQU240.GMPAS-NYF.pm-cpu_gnu.20251003_095213_nlxhjj/bld/cmake-bld/core_ocean/shared/mpas_ocn_tracer_ecosys.f90:2814:57:

 2814 |         if (config_use_ecosysTracers_iron_solubility_file) then
      |                                                         1
Error: Symbol ‘config_use_ecosystracers_iron_solubility_file’ at (1) has no IMPLICIT type
/pscratch/sd/m/mpeterse/e3sm_scratch/pm-cpu/SMS_Ld3.T62_oQU240.GMPAS-NYF.pm-cpu_gnu.20251003_095213_nlxhjj/bld/cmake-bld/core_ocean/shared/mpas_ocn_tracer_ecosys.f90:2464:56:

 2464 |       if (config_use_ecosysTracers_atm_dust_from_coupler .and.  &
      |                                                        1
Error: Symbol ‘config_use_ecosystracers_atm_dust_from_coupler’ at (1) has no IMPLICIT type
/pscratch/sd/m/mpeterse/e3sm_scratch/pm-cpu/SMS_Ld3.T62_oQU240.GMPAS-NYF.pm-cpu_gnu.20251003_095213_nlxhjj/bld/cmake-bld/core_ocean/shared/mpas_ocn_tracer_ecosys.f90:2465:61:

 2465 |           .not. config_use_ecosysTracers_iron_solubility_file) then
      |                                                             1
Error: Symbol ‘config_use_ecosystracers_iron_solubility_file’ at (1) has no IMPLICIT type
```
so something is wrong with the declarations of those variables.

mark-petersen commented:
BTW, on the previous E3SM test, I had to rebase on the head of master to get through the setup on perlmutter. Submitting a job from the head of this PR, there is a library mismatch:
```
Finished SETUP for test SMS_Ld3.T62_oQU240.GMPAS-NYF.pm-cpu_gnu in 2.759769 seconds (FAIL). [COMPLETED 1 of 1]
    Case dir: /pscratch/sd/m/mpeterse/e3sm_scratch/pm-cpu/SMS_Ld3.T62_oQU240.GMPAS-NYF.pm-cpu_gnu.20251003_102449_9pimzr
    Errors were:
        ERROR: module command /usr/share/lmod/lmod/libexec/lmod python load PrgEnv-gnu/8.5.0 gcc-native/12.3 cray-libsci/23.12.5 craype-accel-host craype/2.7.30 cray-mpich/8.1.28 cray-hdf5-parallel/1.12.2.9 cray-netcdf-hdf5parallel/4.9.0.9 cray-parallel-netcdf/1.12.3.9 cmake/3.24.3 evp-patch failed with message:
        Lmod has detected the following error: These module(s) or extension(s) exist but cannot be loaded as requested: "evp-patch"
           Try: "module spider evp-patch" to see how to load the module(s).
```

mark-petersen reviewed (approved):
Never mind. It did pass the smoke test. I've seen this error before, so I'll leave the previous message up there. When you add new config flags, it is best to start with a new repo directory. When I use an old repo directory, there must be some temporary files that remain and then we get errors that the config flags have not been declared.

maltrud committed:
new dust and solubility fields in ocn eco forcing stream

maltrud committed:
define shared constant for iron molecular weight

maltrud commented:
I added the shared constant and updated ocean streams changes and pushed to the branch.  I did a local merge of master with the branch and it passed all except NML since we added new namelist variables.  I think it's ready for the next step.

  SMS_D_Ld1.ne30pg2_r05_IcoswISC30E3r5.WCYCL1850.chrysalis_intel.allactive-wcprod (Overall: NLFAIL) details:
    PASS SMS_D_Ld1.ne30pg2_r05_IcoswISC30E3r5.WCYCL1850.chrysalis_intel.allactive-wcprod CREATE_NEWCASE
    PASS SMS_D_Ld1.ne30pg2_r05_IcoswISC30E3r5.WCYCL1850.chrysalis_intel.allactive-wcprod XML
    PASS SMS_D_Ld1.ne30pg2_r05_IcoswISC30E3r5.WCYCL1850.chrysalis_intel.allactive-wcprod SETUP
    PASS SMS_D_Ld1.ne30pg2_r05_IcoswISC30E3r5.WCYCL1850.chrysalis_intel.allactive-wcprod SHAREDLIB_BUILD time=225
    FAIL SMS_D_Ld1.ne30pg2_r05_IcoswISC30E3r5.WCYCL1850.chrysalis_intel.allactive-wcprod NLCOMP
    PASS SMS_D_Ld1.ne30pg2_r05_IcoswISC30E3r5.WCYCL1850.chrysalis_intel.allactive-wcprod MODEL_BUILD time=275
    PASS SMS_D_Ld1.ne30pg2_r05_IcoswISC30E3r5.WCYCL1850.chrysalis_intel.allactive-wcprod SUBMIT
    PASS SMS_D_Ld1.ne30pg2_r05_IcoswISC30E3r5.WCYCL1850.chrysalis_intel.allactive-wcprod RUN time=1549
    PASS SMS_D_Ld1.ne30pg2_r05_IcoswISC30E3r5.WCYCL1850.chrysalis_intel.allactive-wcprod BASELINE master: ERROR Could not interpret CPRNC output
    PASS SMS_D_Ld1.ne30pg2_r05_IcoswISC30E3r5.WCYCL1850.chrysalis_intel.allactive-wcprod MEMLEAK insufficient data for memleak test
    PASS SMS_D_Ld1.ne30pg2_r05_IcoswISC30E3r5.WCYCL1850.chrysalis_intel.allactive-wcprod SHORT_TERM_ARCHIVER

jonbob committed:
Make bld files consistent with Registry

jonbob commented:
@maltrud -- I ran the semi-automated scripts to make the bld files consistent with what's in Registry. Most of the changes are in ordering of the new configs, but some of the documentation you added to namelist_definition_mpaso.xml will get lost. If any of it is important, please change the descriptions in Registry and I'll rerun the scripts

jonbob commented:
@maltrud -- since this is a stealth feature, we need to add a stealth test with this PR. I'm working on a testdef that just changes the values of these two new configs:
````
    config_use_ecosysTracers_atm_dust_from_coupler
    config_use_ecosysTracers_iron_solubility_file
````
Is that all that needs to be done? Does this need to be a marine BGC run in order for this to work?

rljacob commented:
Update:  working on a test with @njeffery

rljacob commented:
Notes:  needs more work.

njeffery commented:
One more issue that I'm now testing:  If I change the value of dust_FLUX_IN in get_ecosysData for firstTimeStep only because there are no coupling fields yet, then it’s not going to be restartable.  Better to put this in the ocn_import_mct so that dust_FLUX_IN is defined in the get ecosys data only when being read from a file.  Changes to the dust flux or the definition of the iron flux from sea ice coupling are put in the start of the subroutines for internal marbl tendencies/surface fluxes and not in get_ecosysData.

njeffery committed:
Removes flux calculations out of get ecosys data

njeffery committed:
Ensure dust to ocean fluxes are positive

njeffery committed:
Merge branch 'master' into maltrud/ocean/newDustFe

njeffery committed:
Clean up notation after merge

Adds correct dust-iron forcing files to mpaso & mpassi

Changes to sea ice bgc parameters to prevent instabilities in ocean bgc:
1) updated config_mobility_type_dissolved_iron & 2) fixed releaseTime initialization for bgc

njeffery commented:
This is ready to go.   The above changes were tested here: https://e3sm.atlassian.net/wiki/spaces/HESF/pages/6141083649/03232026.v3.LR.HES.PD-dust2.pm-cpu.  

Issues with low surface dissolved iron in the polar regions was resolved by a parameter change in the sea ice

<img width="1676" height="874" alt="Fe_v3 LR HES PD-dust2 pm-cpu_ANN_years0131-0150" src="https://github.com/user-attachments/assets/61fcd5c6-ea4e-4321-8c44-f661505ff3c7" />

@jonbob :  The new ecosys forcing file is here https://web.lcrc.anl.gov/public/e3sm/inputdata/ocn/mpas-o/IcoswISC30E3r5/ecoForcingSurfaceMonthly+WetDryDustSolFrac+CAM6-PD.IcoswISC30E3r5.20251030.nc

njeffery commented:
@proteanplanet :  I've added you to this PR because I had to modify a couple of BGC parameters in the sea ice to keep the ocean bgc from going unstable.  The mpassi changes are BFB in the physics.  The sea ice primary production is significantly decreased in the Southern ocean as expected.

proteanplanet reviewed (commented):
I've been added as a reviewer because this will be included in tag 3.2.  My one charge here is to ensure that this is BFB for LR, HR, SORRM and LRW configurations.  Please can you confirm that to be the case, or else demonstrate the coupled impact if that is not the case.

njeffery commented:
@proteanplanet: This is bfb for all configurations without ocean and sea ice BGC.

proteanplanet reviewed (approved):
Approved for the narrow purpose of inclusion in the V3.2 tag with do-no-harm assurance in place for non-BGC simulations.

rljacob commented:
@njeffery is it ok that this is only in 3.2 (and future tags)?   If you want it available with 3.1, let us know.

njeffery commented:
@rljacob :  3.2 is fine.

njeffery committed:
Adds history output for dust iron

Only modifies streams.ocean when ecosys is active.

jonbob committed:
Remove duplicate variable and fix indenting

jonbob committed:
Go back to the old MARBL tag, update in a separate PR

jonbob committed:
Clean up ocn_comp_mct, move isActive check inside additional logic

jonbob commented:
@maltrud and @njeffery -- please take a peek at the changes I made to ocn_comp_mct. In my testing, I was getting a runtime error for an initialized pointer because we had a check for dust_FLUX_WETField % isActive (and a couple of others) not wrapped in logic for 
````
if ( config_use_ecosysTracers_atm_dust_from_coupler )
````
but I'd feel better if you could review this. The test is resubmitted, and I''ll post if it runs now

jonbob committed:
More ocn_comp_mct cleanup, move isActive check inside additional logic

njeffery reviewed (approved):
Looks great, Jon.  That's the correct logic.

jonbob commented:
With latest updates, testing now shows:
* SMS_D_Ld1.ne30pg2_r05_IcoswISC30E3r5.WCYCL1850.pm-cpu_intel.allactive-wcprod passes with expected NML DIFF
* SMS_Ld2.ne30pg2_r05_IcoswISC30E3r5.BGCEXP_CNTL_CNPECACNT_1850.pm-cpu_intel.elm-bgcexp passes with expected baseline and NML DIFF

jonbob commented:
merged to next

jonbob commented:
merged to master and expected baseline and NML DIFFs blessed
